File: dsa-tracker-backend/src/controllers/authController.js
const User = require('../models/User');
const { generateToken } = require('../utils/jwt');
const crypto = require('crypto');
const sendEmail = require('../utils/sendEmail');

// @desc    Register new user
// @route   POST /api/auth/register
// @access  Public
const registerUser = async (req, res) => {
    try {
        console.log('Register Request Body:', req.body); // DEBUG
        const { name, email, password } = req.body;

        // Check if user exists
        const userExists = await User.findOne({ email });

        if (userExists) {
            return res.status(400).json({ message: 'User already exists' });
        }

        // Create user
        const user = await User.create({
            name,
            email,
            password,
        });

        if (user) {
            res.status(201).json({
                _id: user._id,
                name: user.name,
                email: user.email,
                role: user.role,
                token: generateToken(user._id),
            });
        } else {
            res.status(400).json({ message: 'Invalid user data' });
        }
    } catch (error) {
        console.error('Register Error:', error);

        // Handle Mongoose Validation Error
        if (error.name === 'ValidationError') {
            const message = Object.values(error.errors).map((val) => val.message).join(', ');
            return res.status(400).json({ message });
        }

        // Handle Duplicate Key Error (if not caught by checkout above)
        if (error.code === 11000) {
            return res.status(400).json({ message: 'Email already exists' });
        }

        res.status(500).json({ message: 'Server Error: ' + error.message });
    }
};

// @desc    Login user
// @route   POST /api/auth/login
// @access  Public
const loginUser = async (req, res) => {
    try {
        const { email, password } = req.body;

        // Check for user email
        const user = await User.findOne({ email }).select('+password');

        if (user && (await user.matchPassword(password))) {
            if (user.isBlocked) {
                return res.status(403).json({ message: 'Your account has been blocked. Contact admin.' });
            }

            // Update last active date and streak
            const today = new Date().setHours(0, 0, 0, 0);
            const lastActive = new Date(user.lastActiveDate).setHours(0, 0, 0, 0);
            const daysDiff = Math.floor((today - lastActive) / (1000 * 60 * 60 * 24));

            if (daysDiff === 1) {
                user.streak += 1;
            } else if (daysDiff > 1) {
                user.streak = 1;
            }

            user.lastActiveDate = Date.now();
            await user.save();

            res.json({
                _id: user._id,
                name: user.name,
                email: user.email,
                role: user.role,
                streak: user.streak,
                activeSheet: user.activeSheet,
                hasOnboarded: user.hasOnboarded,
                token: generateToken(user._id),
            });
        } else {
            res.status(401).json({ message: 'Invalid email or password' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Get user profile
// @route   GET /api/auth/profile
// @access  Private
const getUserProfile = async (req, res) => {
    try {
        const user = await User.findById(req.user._id);

        if (user) {
            res.json({
                _id: user._id,
                name: user.name,
                email: user.email,
                role: user.role,
                goal: user.goal,
                dailyGoal: user.dailyGoal,
                streak: user.streak,
                preferredLanguage: user.preferredLanguage,
                xpPoints: user.xpPoints,
                level: user.level,
                activeSheet: user.activeSheet,
                hasOnboarded: user.hasOnboarded,
            });
        } else {
            res.status(404).json({ message: 'User not found' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Update user profile
// @route   PUT /api/auth/profile
// @access  Private
const updateUserProfile = async (req, res) => {
    try {
        const user = await User.findById(req.user._id);

        if (user) {
            user.name = req.body.name || user.name;
            user.goal = req.body.goal || user.goal;
            user.dailyGoal = req.body.dailyGoal || user.dailyGoal;
            user.preferredLanguage = req.body.preferredLanguage || user.preferredLanguage;

            // Handle activeSheet update (allow null for Overall)
            if (req.body.activeSheet !== undefined) {
                user.activeSheet = req.body.activeSheet;
            }

            if (req.body.hasOnboarded !== undefined) {
                user.hasOnboarded = req.body.hasOnboarded;
            }

            const updatedUser = await user.save();

            res.json({
                _id: updatedUser._id,
                name: updatedUser.name,
                email: updatedUser.email,
                goal: updatedUser.goal,
                dailyGoal: updatedUser.dailyGoal,
                preferredLanguage: updatedUser.preferredLanguage,
                activeSheet: updatedUser.activeSheet,
                hasOnboarded: updatedUser.hasOnboarded,
            });
        } else {
            res.status(404).json({ message: 'User not found' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Google auth callback
// @route   GET /api/auth/google/callback
// @access  Public
const googleCallback = (req, res) => {
    const token = generateToken(req.user._id);
    const frontendUrl = process.env.FRONTEND_URL || 'http://localhost:3000';
    // Redirect to frontend with token
    res.redirect(`${frontendUrl}/oauth/callback?token=${token}`);
};

// @desc    Update user password
// @route   PUT /api/auth/updatepassword
// @access  Private
const updatePassword = async (req, res) => {
    try {
        const user = await User.findById(req.user._id).select('+password');

        if (user) {
            // Check current password
            if (await user.matchPassword(req.body.currentPassword)) {
                user.password = req.body.newPassword;
                await user.save();
                res.json({ message: 'Password updated' });
            } else {
                res.status(401).json({ message: 'Invalid current password' });
            }
        } else {
            res.status(404).json({ message: 'User not found' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Forgot Password
// @route   POST /api/auth/forgotpassword
// @access  Public
const forgotPassword = async (req, res) => {
    console.log('Forgot Password Hit. Body:', req.body); // DEBUG
    try {
        const user = await User.findOne({ email: req.body.email });

        if (!user) {
            return res.status(404).json({ message: 'User not found' });
        }

        // Generate 6 digit OTP
        const otp = Math.floor(100000 + Math.random() * 900000).toString();

        // Hash token and set to resetPasswordToken field (using same field for OTP hash)
        user.resetPasswordToken = crypto.createHash('sha256').update(otp).digest('hex');
        user.resetPasswordExpire = Date.now() + 10 * 60 * 1000; // 10 minutes

        await user.save({ validateBeforeSave: false });

        const message = `Your password reset OTP is: ${otp} \n\n It is valid for 10 minutes. \n\n If you didn't request this, please ignore this email. \n\n Please check your spam folder if you don't see this email in your inbox.`;

        console.log('Attempting to send email to:', user.email);
        console.log('SMTP_HOST:', process.env.SMTP_HOST);
        console.log('SMTP_USER:', process.env.SMTP_EMAIL);
        // Do not log password


        try {
            await sendEmail({
                email: user.email,
                subject: 'DSA Tracker - Password Reset OTP',
                message,
            });

            res.status(200).json({ message: 'OTP sent to email. Please check your spam folder too.' });
        } catch (error) {
            console.error(error);
            const fs = require('fs');
            fs.appendFileSync('email_error.log', new Date().toISOString() + ' - ' + error.message + '\n' + JSON.stringify(error) + '\n');

            user.resetPasswordToken = undefined;
            user.resetPasswordExpire = undefined;

            await user.save({ validateBeforeSave: false });

            return res.status(500).json({ message: 'Email could not be sent' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Reset Password
// @route   POST /api/auth/resetpassword/verify
// @access  Public
const resetPassword = async (req, res) => {
    try {
        const { email, otp, password } = req.body;

        if (!email || !otp || !password) {
            return res.status(400).json({ message: 'Please provide email, OTP and new password' });
        }

        // Get hashed token
        const resetPasswordToken = crypto
            .createHash('sha256')
            .update(otp)
            .digest('hex');

        const user = await User.findOne({
            email,
            resetPasswordToken,
            resetPasswordExpire: { $gt: Date.now() },
        });

        if (!user) {
            return res.status(400).json({ message: 'Invalid or expired OTP' });
        }

        // Set new password
        user.password = password;
        user.resetPasswordToken = undefined;
        user.resetPasswordExpire = undefined;
        await user.save();

        res.status(200).json({
            message: 'Password reset successful',
            token: generateToken(user._id),
        });
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

module.exports = {
    registerUser,
    loginUser,
    getUserProfile,
    updateUserProfile,
    googleCallback,
    updatePassword,
    forgotPassword,
    resetPassword,
};

File: dsa-tracker-backend/src/main/java/com/dsatracker/controller/AuthController.java
package com.dsatracker.controller;

import com.dsatracker.dto.*;
import com.dsatracker.exception.BadRequestException;
import com.dsatracker.exception.ResourceNotFoundException;
import com.dsatracker.model.User;
import com.dsatracker.repository.UserRepository;
import com.dsatracker.security.JwtTokenProvider;
import com.dsatracker.service.EmailService;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpEntity;
import org.springframework.http.MediaType;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.*;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider tokenProvider;

    @Autowired
    private EmailService emailService;

    @Value("${google.client.id:your_google_client_id}")
    private String googleClientId;

    @Value("${google.client.secret:your_google_client_secret}")
    private String googleClientSecret;

    @Value("${google.redirect.uri:http://localhost:5000/api/auth/google/callback}")
    private String googleRedirectUri;

    @Value("${frontend.url:http://localhost:3000}")
    private String frontendUrl;

    // Helper to get authenticated user
    private User getAuthenticatedUser() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal instanceof UserDetails) {
            String userId = ((UserDetails) principal).getUsername();
            return userRepository.findById(userId)
                    .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        }
        throw new BadRequestException("Unauthorized access");
    }

    private String hashOtp(String otp) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(otp.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Error hashing OTP", e);
        }
    }

    @PostMapping("/register")
    public ResponseEntity<?> registerUser(@Valid @RequestBody RegisterRequest registerRequest) {
        if (userRepository.findByEmail(registerRequest.getEmail()).isPresent()) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "User already exists");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        User user = User.builder()
                .name(registerRequest.getName())
                .email(registerRequest.getEmail().toLowerCase())
                .password(passwordEncoder.encode(registerRequest.getPassword()))
                .build();

        User savedUser = userRepository.save(user);

        String token = tokenProvider.generateToken(savedUser.getId());

        AuthResponse authResponse = AuthResponse.builder()
                .token(token)
                ._id(savedUser.getId())
                .name(savedUser.getName())
                .email(savedUser.getEmail())
                .role(savedUser.getRole())
                .streak(savedUser.getStreak())
                .activeSheet(savedUser.getActiveSheet())
                .hasOnboarded(savedUser.isHasOnboarded())
                .build();

        return ResponseEntity.status(HttpStatus.CREATED).body(authResponse);
    }

    @PostMapping("/login")
    public ResponseEntity<?> loginUser(@Valid @RequestBody LoginRequest loginRequest) {
        User user = userRepository.findByEmail(loginRequest.getEmail().toLowerCase())
                .orElseThrow(() -> new BadRequestException("Invalid email or password"));

        if (!passwordEncoder.matches(loginRequest.getPassword(), user.getPassword())) {
            throw new BadRequestException("Invalid email or password");
        }

        if (user.isBlocked()) {
            Map<String, String> response = new HashMap<>();
            response.put("message", "Your account has been blocked. Contact admin.");
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(response);
        }

        // Update streak and last active date
        Calendar todayCal = Calendar.getInstance();
        todayCal.set(Calendar.HOUR_OF_DAY, 0);
        todayCal.set(Calendar.MINUTE, 0);
        todayCal.set(Calendar.SECOND, 0);
        todayCal.set(Calendar.MILLISECOND, 0);
        long today = todayCal.getTimeInMillis();

        Calendar activeCal = Calendar.getInstance();
        activeCal.setTime(user.getLastActiveDate());
        activeCal.set(Calendar.HOUR_OF_DAY, 0);
        activeCal.set(Calendar.MINUTE, 0);
        activeCal.set(Calendar.SECOND, 0);
        activeCal.set(Calendar.MILLISECOND, 0);
        long lastActive = activeCal.getTimeInMillis();

        long diffMs = today - lastActive;
        long daysDiff = diffMs / (1000 * 60 * 60 * 24);

        if (daysDiff == 1) {
            user.setStreak(user.getStreak() + 1);
        } else if (daysDiff > 1) {
            user.setStreak(1);
        }

        user.setLastActiveDate(new Date());
        userRepository.save(user);

        String token = tokenProvider.generateToken(user.getId());

        AuthResponse authResponse = AuthResponse.builder()
                .token(token)
                ._id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .streak(user.getStreak())
                .activeSheet(user.getActiveSheet())
                .hasOnboarded(user.isHasOnboarded())
                .build();

        return ResponseEntity.ok(authResponse);
    }

    @GetMapping("/profile")
    public ResponseEntity<?> getUserProfile() {
        User user = getAuthenticatedUser();
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("_id", user.getId());
        response.put("name", user.getName());
        response.put("email", user.getEmail());
        response.put("role", user.getRole());
        response.put("goal", user.getGoal());
        response.put("dailyGoal", user.getDailyGoal());
        response.put("streak", user.getStreak());
        response.put("preferredLanguage", user.getPreferredLanguage());
        response.put("xpPoints", user.getXpPoints());
        response.put("level", user.getLevel());
        response.put("activeSheet", user.getActiveSheet());
        response.put("hasOnboarded", user.isHasOnboarded());
        return ResponseEntity.ok(response);
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateUserProfile(@RequestBody UpdateProfileRequest updateProfileRequest) {
        User user = getAuthenticatedUser();

        if (updateProfileRequest.getName() != null) user.setName(updateProfileRequest.getName());
        if (updateProfileRequest.getGoal() != null) user.setGoal(updateProfileRequest.getGoal());
        if (updateProfileRequest.getDailyGoal() != null) user.setDailyGoal(updateProfileRequest.getDailyGoal());
        if (updateProfileRequest.getPreferredLanguage() != null) user.setPreferredLanguage(updateProfileRequest.getPreferredLanguage());
        if (updateProfileRequest.getActiveSheet() != null) {
            String activeSheet = updateProfileRequest.getActiveSheet();
            if (!org.springframework.util.StringUtils.hasText(activeSheet) || "null".equalsIgnoreCase(activeSheet)) {
                user.setActiveSheet(null);
            } else {
                user.setActiveSheet(activeSheet);
            }
        }
        if (updateProfileRequest.getHasOnboarded() != null) user.setHasOnboarded(updateProfileRequest.getHasOnboarded());

        User updatedUser = userRepository.save(user);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("_id", updatedUser.getId());
        response.put("name", updatedUser.getName());
        response.put("email", updatedUser.getEmail());
        response.put("goal", updatedUser.getGoal());
        response.put("dailyGoal", updatedUser.getDailyGoal());
        response.put("preferredLanguage", updatedUser.getPreferredLanguage());
        response.put("activeSheet", updatedUser.getActiveSheet());
        response.put("hasOnboarded", updatedUser.isHasOnboarded());

        return ResponseEntity.ok(response);
    }



    @PutMapping("/updatepassword")
    public ResponseEntity<?> updatePassword(@RequestBody Map<String, String> body) {
        String currentPassword = body.get("currentPassword");
        String newPassword = body.get("newPassword");

        if (!StringUtils.hasText(currentPassword) || !StringUtils.hasText(newPassword)) {
            throw new BadRequestException("Please provide current and new password");
        }

        User user = getAuthenticatedUser();

        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new BadRequestException("Invalid current password");
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Password updated");
        return ResponseEntity.ok(response);
    }

    @PostMapping("/forgotpassword")
    public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        if (!StringUtils.hasText(email)) {
            throw new BadRequestException("Please provide email");
        }

        User user = userRepository.findByEmail(email.toLowerCase())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        // Generate 6 digit OTP
        SecureRandom random = new SecureRandom();
        int otpCode = 100000 + random.nextInt(900000);
        String otp = String.valueOf(otpCode);

        // Save hashed OTP and expiration in User Document
        user.setResetPasswordToken(hashOtp(otp));
        // Expiration = 10 mins
        user.setResetPasswordExpire(new Date(System.currentTimeMillis() + 10 * 60 * 1000));
        userRepository.save(user);

        String emailText = "Your password reset OTP is: " + otp + 
                "\n\nIt is valid for 10 minutes. \n\nIf you didn't request this, please ignore this email.";

        emailService.sendEmail(user.getEmail(), "DSA Tracker - Password Reset OTP", emailText);

        Map<String, String> response = new HashMap<>();
        response.put("message", "OTP sent to email. Please check your spam folder too.");
        return ResponseEntity.ok(response);
    }

    @PostMapping("/resetpassword/verify")
    public ResponseEntity<?> resetPassword(@Valid @RequestBody ResetPasswordRequest resetPasswordRequest) {
        User user = userRepository.findByEmail(resetPasswordRequest.getEmail().toLowerCase())
                .orElseThrow(() -> new BadRequestException("Invalid request"));

        String hashedOtp = hashOtp(resetPasswordRequest.getOtp());

        if (user.getResetPasswordToken() == null || 
            !user.getResetPasswordToken().equals(hashedOtp) || 
            user.getResetPasswordExpire() == null || 
            user.getResetPasswordExpire().before(new Date())) {
            throw new BadRequestException("Invalid or expired OTP");
        }

        user.setPassword(passwordEncoder.encode(resetPasswordRequest.getPassword()));
        user.setResetPasswordToken(null);
        user.setResetPasswordExpire(null);
        userRepository.save(user);

        String token = tokenProvider.generateToken(user.getId());

        Map<String, String> response = new HashMap<>();
        response.put("message", "Password reset successful");
        response.put("token", token);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/google")
    public void googleLogin(HttpServletResponse response) throws IOException {
        String googleAuthUrl = "https://accounts.google.com/o/oauth2/v2/auth" +
                "?client_id=" + googleClientId +
                "&redirect_uri=" + googleRedirectUri +
                "&response_type=code" +
                "&scope=profile%20email";
        response.sendRedirect(googleAuthUrl);
    }

    @GetMapping("/google/callback")
    public void googleCallback(@RequestParam("code") String code, HttpServletResponse response) throws IOException {
        try {
            RestTemplate restTemplate = new RestTemplate();
            String tokenUrl = "https://oauth2.googleapis.com/token";

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

            MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
            map.add("code", code);
            map.add("client_id", googleClientId);
            map.add("client_secret", googleClientSecret);
            map.add("redirect_uri", googleRedirectUri);
            map.add("grant_type", "authorization_code");

            HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(map, headers);
            Map<String, Object> tokenResponse = restTemplate.postForObject(tokenUrl, requestEntity, Map.class);

            if (tokenResponse == null || !tokenResponse.containsKey("access_token")) {
                throw new BadRequestException("Failed to retrieve access token from Google");
            }

            String accessToken = (String) tokenResponse.get("access_token");

            String userInfoUrl = "https://www.googleapis.com/oauth2/v3/userinfo?access_token=" + accessToken;
            Map<String, Object> userInfo = restTemplate.getForObject(userInfoUrl, Map.class);

            if (userInfo == null || !userInfo.containsKey("email")) {
                throw new BadRequestException("Failed to retrieve user info from Google");
            }

            String googleId = (String) userInfo.get("sub");
            String email = ((String) userInfo.get("email")).toLowerCase();
            String name = (String) userInfo.get("name");

            User user = userRepository.findByGoogleId(googleId).orElse(null);

            if (user == null) {
                user = userRepository.findByEmail(email).orElse(null);
                if (user != null) {
                    user.setGoogleId(googleId);
                    userRepository.save(user);
                } else {
                    user = User.builder()
                            .name(name)
                            .email(email)
                            .googleId(googleId)
                            .password(passwordEncoder.encode(UUID.randomUUID().toString()))
                            .build();
                    user = userRepository.save(user);
                }
            }

            String token = tokenProvider.generateToken(user.getId());
            response.sendRedirect(frontendUrl + "/oauth/callback?token=" + token);

        } catch (Exception e) {
            response.sendRedirect(frontendUrl + "/login?error=Google authentication failed: " + e.getMessage());
        }
    }
}
